go/tasks/plugins/k8s/sagemaker/utils.go

```diff
@@ -0,0 +1,281 @@
+			}
+			retValue.ContinuousParameterRanges = append(retValue.ContinuousParameterRanges, newElem)
+
+		case *sagemakerSpec.ParameterRangeOneOf_IntegerParameterRange:
+			var newElem = commonv1.IntegerParameterRange{
+				MaxValue:    awssagemaker.ToStringPtr(fmt.Sprintf("%d", pr.GetIntegerParameterRange().GetMaxValue())),
+				MinValue:    awssagemaker.ToStringPtr(fmt.Sprintf("%d", pr.GetIntegerParameterRange().GetMinValue())),
+				Name:        awssagemaker.ToStringPtr(prName),
+				ScalingType: getAPIScalingType(pr.GetContinuousParameterRange().GetScalingType()),
+			}
+			retValue.IntegerParameterRanges = append(retValue.IntegerParameterRanges, newElem)
+		}
+	}
+
+	return retValue
+}
+
+func convertHyperparameterTuningJobConfigToSpecType(hpoJobConfigLiteral *core.Literal) (*sagemakerSpec.HyperparameterTuningJobConfig, error) {
+	var retValue = &sagemakerSpec.HyperparameterTuningJobConfig{}
+	hpoJobConfigByteArray := hpoJobConfigLiteral.GetScalar().GetBinary().GetValue()
+	err := proto.Unmarshal(hpoJobConfigByteArray, retValue)
+	if err != nil {
+		return nil, errors.Errorf("Hyperparameter Tuning Job Config Literal in input cannot be unmarshalled into spec type")
+	}
+	return retValue, nil
+}
+
+func convertStaticHyperparamsLiteralToSpecType(hyperparamLiteral *core.Literal) ([]*commonv1.KeyValuePair, error) {
+	var retValue []*commonv1.KeyValuePair
+	hyperFields := hyperparamLiteral.GetScalar().GetGeneric().GetFields()
+	if hyperFields == nil {
+		return nil, errors.Errorf("Failed to get the static hyperparameters field from the literal")
+	}
+	for k, v := range hyperFields {
+		var newElem = commonv1.KeyValuePair{
+			Name:  k,
+			Value: v.GetStringValue(),
+		}
+		retValue = append(retValue, &newElem)
+	}
+	return retValue, nil
+}
+
+func ToStringPtr(str string) *string {
+	if str == "" {
+		return nil
+	}
+	return &str
+}
+
+func ToInt64Ptr(i int64) *int64 {
+	if i == 0 {
+		return nil
+	}
+	return &i
+}
+
+func ToIntPtr(i int) *int {
+	if i == 0 {
+		return nil
+	}
+	return &i
+}
+
+func ToFloat64Ptr(f float64) *float64 {
+	if f == 0 {
+		return nil
+	}
+	return &f
+}
+
+func createOutputLiteralMap(tk *core.TaskTemplate, outputPath string) *core.LiteralMap {
+	op := &core.LiteralMap{}
+	for k := range tk.Interface.Outputs.Variables {
+		// if v != core.LiteralType_Blob{}
+		op.Literals = make(map[string]*core.Literal)
+		op.Literals[k] = &core.Literal{
+			Value: &core.Literal_Scalar{
+				Scalar: &core.Scalar{
+					Value: &core.Scalar_Blob{
+						Blob: &core.Blob{
+							Metadata: &core.BlobMetadata{
+								Type: &core.BlobType{Dimensionality: core.BlobType_SINGLE},
+							},
+							Uri: outputPath,
+						},
+					},
+				},
+			},
+		}
+	}
+	return op
+}
+
+func deleteConflictingStaticHyperparameters(
+	ctx context.Context,
+	staticHPs []*commonv1.KeyValuePair,
+	tunableHPMap map[string]*sagemakerSpec.ParameterRangeOneOf) []*commonv1.KeyValuePair {
+
+	w := 0 // write position
```

Comment:
why do you need this, why not just use append? Anyways you are creating a new array. I think you just need to change `make([]type, 0, len(...))`
What this will do is create a slice of length `len` but start at position 0. So now you can just append

Reply:
Isn't it going to create new slice whenever `append` is called?

Reply:
Are these two approaches incurring the same number of memory block re-allocations? but yeah your suggestion is more elegant